README.md
# ipayment-rails

Ruby on Rails implementation for the european 1&1 iPayment gateway for credit card processing.

* [Project page](http://saulabs.net/projects/ipayment-rails)
* [Repository](http://github.com/saulabs/ipayment-rails)

Technical documentation from iPayment:
[http://www.1und1.info/downloads/ipayment_Technik-Handbuch_2008-08.pdf](http://www.1und1.info/downloads/ipayment_Technik-Handbuch_2008-08.pdf) (German)

This plugin provides a simple way to implement credit card processing by providing wrappers and helpers around the 1&1 iPayment SOAP API. As the credit card data doesn't touch your application's servers, **no PCI certification is needed**.

## Examples

See the _example/_ directory for some example code.

## How it works (single payment)

* *form_for_ipayment_auth* creates a session and stores transaction data into it (amount, currency, etc.).
* The HTML form POSTs the user entered credit card data SSL secured directly to the iPayment servers.
* The iPayment servers will process the transaction and request the *callback_url* provided to *form_for_ipayment_auth* passing the result to it. At this point your application should handle the transaction, e.g. booking the amount to the user account if it is successful.
* iPayment will redirect the user to *success_url* if the transactions succeeds or to *error_url* if it doesn't.
* All this happens transparently to the user – except for a short change in the address field of the browser, he stays on your site.

## How it works (recurring payments)

* *form_for_ipayment_check_save* creates a session and stores transaction data into it (options, etc.).
* The HTML form POSTs the user entered credit card data SSL secured directly to the iPayment servers.
* The iPayment servers check the credit card but don't charge it and request the *callback_url* provided to *form_for_ipayment_check_save* passing the result with an transaction id to it. At this point your application should handle the transaction, e.g. create a subscription for the user, storing the transaction id for the following charges.
* iPayment will redirect the user to *success_url* if the check succeeds or to *error_url* if it doesn't.
* All this happens transparently to the user – except for a short change in the address field of the browser, he stays on your site.
* For the first payment (e.g after a month of trial), create an Ipayment::Payment and set it's *initial_recurring* attribute to true and charge the amount with *charge!(:orig_trx_number => your_initial_transcation_number)*
* For the second and following payments payment, create an Ipayment::Payment and set it's *recurring* attribute to true and charge the amount with *charge!(:orig_trx_number => your_initial_transcation_number)*

## How to get it work

* Get iPayment credentials and put them into the config file. The file goes into *RAILS_ROOT/config/ipayment.yml* (or use a test account, see *ipayment_Technik-Handbuch_2008-08.pdf* page 10).
* Implement a controller to handle callback, success and error (see controller.rb example).
* Add a view using *form_for_ipayment_auth* passing it an Ipayment::Payment with payment data (see controller.rb and new.html.erb).

## Note on Patches/Pull Requests
 
* Fork the project.
* Make your feature addition or bug fix.
* Add specs for it. This is important so I don't break it in a
  future version unintentionally.
* Commit, do not mess with rakefile, version, or history.
  (if you want to have your own version, that is fine but
   bump version in a commit by itself I can ignore when I pull)
* Send me a pull request. Bonus points for topic branches.

## Copyright

Copyright (c) 2009-2010 Dieter Komendera. See LICENSE for details.


CHANGELOG
*0.2*

* Ruby version 1.9.2 required
* Changed the way how to pass configuration parameters to support multiple configurations
* Added support for recurring payments
* Fixed encoding issues


*0.1*

* First release


spec/spec_helper.rb
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))
require 'rspec'
require 'ipayment'

# Requires supporting files with custom matchers and macros, etc,
# in ./support/ and its subdirectories.
Dir["#{File.dirname(__FILE__)}/support/**/*.rb"].each {|f| require f}

RSpec.configure do |config|
  
end


spec/ipayment_spec.rb
# -*- encoding : utf-8 -*-
require File.expand_path(File.dirname(__FILE__) + '/spec_helper')

describe Ipayment do
  
  class Rails
    class Env
      def self.development?
        false
      end
    end
    def self.env
      Env
    end
  end

  describe 'Connection' do
    it "should initialise a driver when initialized" do
      factory = mock('factory')
      factory.should_receive(:create_rpc_driver).once.and_return(driver = mock('driver', :wiredump_file_base= => nil))
      SOAP::WSDLDriverFactory.should_receive(:new).and_return(factory)
      connection = Ipayment::Connection.new
      connection.driver.should == driver
    end
  end

  describe 'Helpers module' do
    class MyViewClass
      include Ipayment::Helpers
      
      def protect_against_forgery?
        false
      end
      
    end

    before(:each) do
      @ipayment = mock('ipayment', :generate_session_id => '1234', :config => ipayment_config)
      @view = MyViewClass.new
      @view.stub!(:concat).and_return('concatinated')
      @view.stub!(:hidden_field_tag).and_return('hidden_field_tag')
      @view.stub!(:form_tag).and_yield().and_return('form_tag')
    end

    it "should generate session id for the ipayment" do
      @ipayment.should_receive(:generate_session_id).with({
        :success_url => 'success_url', :error_url => 'success_url',
        :transaction_type => 'auth'
      }).and_return('1234')
      @view.form_for_ipayment_auth(@ipayment, { :success_url => 'success_url',
        :error_url => 'success_url' }) {}
    end

    it "should generate a form" do
      @view.should_receive(:form_tag).with(
        "https://ipayment.de/merchant/#{@ipayment.config['accountId']}/processor/2.0/",
        { :class => 'cc-form' })
      @view.form_for_ipayment_auth(@ipayment, { :success_url => 'success_url',
        :error_url => 'success_url', :class => 'cc-form' }) {}
    end

    it "should generate 3 hidden input fields" do
      # FIXME: we should test if the hidden inputs get the right arguments, no idea how to implemtn that with rspec
      @view.should_receive(:hidden_field_tag).exactly(4).times.and_return('hidden_field_tag')
      @view.form_for_ipayment_auth(@ipayment, { :success_url => 'success_url',
        :error_url => 'success_url', :callback_url => 'callback_url',
        :from_datastorage_id => '987', :class => 'cc-form' }) {}
    end
  end

  describe 'Controller module' do
    class MyPaymentsController
      attr_accessor :request
      include Ipayment::Controller
    end

    before(:each) do
      @controller = MyPaymentsController.new
    end

    describe "validate_ipayment_request_source" do
      it "should be valid when request is post and ip is correct" do
        request = mock('request')
        request.should_receive(:env).and_return({ 'REMOTE_ADDR' => '212.227.34.218' })
        request.should_receive(:post?).and_return(true)
        @controller.request = request
        @controller.send(:validate_ipayment_request_source).should be_true
      end

      it "should be invalid when request is not post and ip is correct" do
        request = mock('request')
        request.should_receive(:post?).and_return(false)
        @controller.request = request
        @controller.send(:validate_ipayment_request_source).should be_false
      end

      it "should be invalid when request is post and ip is incorrect" do
        request = mock('request')
        request.should_receive(:env).and_return({ 'REMOTE_ADDR' => '212.227.34.100' })
        request.should_receive(:post?).and_return(true)
        @controller.request = request
        @controller.send(:validate_ipayment_request_source).should be_false
      end
    end

    describe "handle_ipayment_callback" do

      it "should validate request source" do
        @controller.should_receive(:validate_ipayment_request_source).and_return(false)
        @controller.send(:handle_ipayment_callback, ipayment_config) { |x| }
      end

      it "should yield nil if request source is not valid" do
        @controller.stub!(:validate_ipayment_request_source).and_return(false)
        @controller.send(:handle_ipayment_callback, ipayment_config) do |payment|
          payment.should be_nil
        end
      end

      it "parse Ipayment::Payment form params" do
        @controller.stub!(:validate_ipayment_request_source).and_return(true)
        @controller.should_receive(:params).and_return(params = { :id => 1 })
        Ipayment::Payment.should_receive(:parse_from_params).with(ipayment_config, params).and_return(payment = mock('ipayment'))
        @controller.send(:handle_ipayment_callback, ipayment_config) do |payment|
          payment.should == payment
        end
      end
    end
  end

  describe 'Service' do
    before(:each) do
      @driver = mock('driver')
      @service =  Ipayment::Service.new(ipayment_config)
      @service.stub!(:driver).and_return(@driver)
      @service.stub!(:account_data).and_return(@account_data = { 'accountId' => '123' })
    end

    describe "create_session" do
      it "should map params correctly for the driver call" do
        @driver.should_receive(:createSession).with(@account_data,
          { 'trxAmount' => '100', 'trxCurrency' => 'EUR', 'invoiceText' => 'Invoice1' },
          'auth', 'cc', {}, { 'redirectUrl' => 'success', 'silentErrorUrl' => 'error' }
        ).and_return(['1234'])
        @service.create_session({
          :amount => '100', :currency => 'EUR', :payment_type => 'cc', :options => {},
          :transaction_type => 'auth', :success_url => 'success', :error_url => 'error',
          :invoice_text => 'Invoice1'
        }).should == '1234'
      end
    end
  end

  describe 'Payment' do

    describe 'initialize' do
      it "should populate attributes on initialize" do
        payment = Ipayment::Payment.new(ipayment_config, payment_attributes)
        payment.amount.should == 100
        payment.currency.should == 'EUR'
        payment.payment_type.should == 'cc'
        payment.options.should == { :test => 1 }
      end

      it "should raise error if initialized with unknown attribute" do
        lambda {
          Ipayment::Payment.new(ipayment_config, { :fuu => 100 })
        }.should raise_error
      end
    end

    describe 'generate_session_id' do
      it "should create session with own attributes merged into params" do
        @service = Ipayment::Service.new(ipayment_config)
        @service.should_receive(:create_session, payment_attributes(:fuu => 'bar')).and_return('12334')
        payment = Ipayment::Payment.new(ipayment_config, payment_attributes)
        payment.instance_variable_set("@service", @service)
        payment.generate_session_id(:fuu => 'bar').should == '12334'
      end
    end

    describe 'parse_from_params' do
      before(:all) do
        @payment = Ipayment::Payment.parse_from_params(ipayment_config, params)
      end

      it "should set correct attributes" do
        @payment.amount.should == '2000'
        @payment.currency.should == 'EUR'
        @payment.payment_type.should == 'cc'
        @payment.options.should == params
      end

      describe "on success" do
        it "should set @paid to true if response was success and transaction type auth" do
          @payment.paid?.should be_true
        end

        it "should set @paid to true if resonse was success but transaction type was not auth" do
          @payment = Ipayment::Payment.parse_from_params(ipayment_config, params('trx_typ' => 'capture'))
          @payment.paid?.should be_false
        end

        it "should set transaction_number" do
          @payment.transaction_number.should == "1-30546206"
        end

        it "should parse cc data" do
          @payment.cc_data.should == {
            :number       => "XXXXXXXXXXXX1111",
            :card_type    => "VisaCard",
            :name         => "Hans Huber",
            :expiry_month => "01",
            :expiry_year  =>  "10"
          }
        end

        it "should parse error data" do
          @payment.error_code.should be_zero
          @payment.error_message.should be_nil
        end

        it "should parse cc data with umlauts" do
          # iPayment sends LATIN1
          @payment = Ipayment::Payment.parse_from_params(ipayment_config, params('paydata_cc_cardowner' => Iconv.conv('LATIN1', 'UTF8', 'HEINZ MÜLLER')))
          @payment.cc_data[:name].should == "HEINZ MÜLLER"
        end
      end

      describe "on error" do
        before(:all) do
          @payment = Ipayment::Payment.parse_from_params(ipayment_config, params_error)
        end

        it "should set @paid to false if resonse was no success" do
          @payment.paid?.should be_false
        end

        it "should parse error data" do
          @payment.error_code.should == 5002
          @payment.error_message.should == 'Die angegebene Kreditkartennummer ist fehlerhaft.'
        end
      end
    end

    def payment_attributes(options = {})
      { :amount => 100, :currency => 'EUR', :invoice_text => 'Invoice123',
        :payment_type => 'cc', :options => { :test => 1 } }.merge(options)
    end

    def params(options = {})
      # typical params from a successful callback request
      {
        "trx_amount"=>"2000", "trx_remoteip_country"=>"NZ", "storage_id"=>"5068698",
        "commit"=>"Abschicken", "trxuser_id"=>"99999", "trx_paymenttyp"=>"cc", "trx_currency"=>"EUR",
        "ret_transtime"=>"07:26:33", "ret_trx_number"=>"1-30546206", "action"=>"ipayment",
        "ret_transdate"=>"27.03.09", "ret_booknr"=>"1-30546206", "trx_paymentmethod"=>"VisaCard",
        "trx_paymentdata_country"=>"US", "id"=>"callback", "addr_name"=>"Hans Huber",
        "ret_errorcode"=>"0", "paydata_cc_number"=>"XXXXXXXXXXXX1111", "paydata_cc_typ"=>"VisaCard",
        "ret_status"=>"SUCCESS", "redirect_needed"=>"0", "paydata_cc_expdate"=>"0110",
        "controller"=>"payments", "trx_typ"=>"auth", "paydata_cc_cardowner"=>"Hans Huber",
        "payment_id"=>"14496", "ret_ip"=>"202.20.7.150", "ret_authcode"=>""
      }.merge(options)
    end

    def params_error(options = {})
      # typical params for a error callback request
      {
        "trx_amount"=>"3000", "commit"=>"Abschicken", "invoice_text"=>"CC4D77E8600",
        "ret_fatalerror"=>"0", "action"=>"error", "trx_currency"=>"EUR", "ret_errorcode"=>"5002",
        "ret_errormsg"=>"Die angegebene Kreditkartennummer ist fehlerhaft.", "ret_status"=>"ERROR",
        "controller"=>"payments/ipayment", "trx_typ"=>"auth",
        "ret_additionalmsg"=>"The Creditcard-Number is invalid (0)", "redirect_needed"=>"0",
        "addr_name"=>"Hans Test", "payment_id"=>"14497", "ret_ip"=>"202.20.7.150",
        "trxuser_id"=>"9491", "trx_paymenttyp"=>"cc"
      }.merge(options)
    end
  end

  def ipayment_config
    {
      'accountId' => 99999,
      'trxuserId' =>  99999,
      'trxpassword' => 0,
      'adminactionpassword' => '5cfgRT34xsdedtFLdfHxj7tfwx24fe'
    }
  end

end

if ENV['LIVE_TESTS'] == '1'
  describe Ipayment, "live" do
    
    before(:each) do
      @service = Ipayment::Service.new(ipayment_config)
    end
    
    def test_storage_id
      "8086202"
    end

    describe "createSession" do

      it "should return a valid session id" do
        @service.create_session({
          :amount => 100,
          :currency => 'EUR',
          :type => 'auth',
          :payment_type => 'cc',
          :transaction_type => "auth",
          :redirect_url => 'http://sauspiel.local/success',
          :silent_error_url => 'http://sauspiel.local/fail'
        }).should match(/[a-zA-Z0-9]{32}/)
      end
    end

    def ipayment_config
      {
        'accountId' => 99999,
        'trxuserId' =>  99999,
        'trxpassword' => 0,
        'adminactionpassword' => '5cfgRT34xsdedtFLdfHxj7tfwx24fe'
      }
    end
  end

end



spec/ruby-ipayment_spec.rb
require File.expand_path(File.dirname(__FILE__) + '/spec_helper')

describe "RubyIpayment" do
  it "fails" do
    fail "hey buddy, you should probably rename this file and start specing for real"
  end
end


example/controller.rb
class PaymentController < ApplicationController
  helper Ipayment::Helpers
  include Ipayment::Controller

  def new
    @ipayment = Ipayment::Payment.new({
      :amount => 1000,
      :currency => 'EUR',
      :payment_type => 'cc',
      :invoice_text => 'Payment #1234',
      :options => {
        :payment_id => 1111 # e.g your internal payment handle
      }
    })
  end

  def callback
    # requested from the ipayment servers
    handle_ipayment_callback do |ipayment|
      # ipayment parameter is an Ipayment::Payment instance
      # This is the right place to handle a payment, e.g credit the payment amount to the user.
      # You've access to the options passed to the Ipayment::Payment instance, e.g:
      # ipayment.options['payment_id'] #=> 1111
    end
    head :ok
  end

  def success
    # user gets redirected here if transaction succeeds
    # the options, in this example the payment_id get passed in the params
    # e.g params[:payment_id] #=> 1111
  end

  def error
    # user gets redirected here if transaction fails
    @ipayment = Ipayment::Payment.parse_from_params(params)
    if @ipayment && @ipayment.error_code != 0
      @error = @ipayment.error_message
    elsif params[:internal] == 'true'
      @error = "An error occured while processing the transaction"
    end
  end
end


example/single/payments_controller.rb
class PaymentController < ApplicationController
  helper Ipayment::Helpers
  include Ipayment::Controller

  before_filter :set_ipayment_config

  def new
    @ipayment = Ipayment::Payment.new(@ipayment_config, {
      :amount => 1000,
      :currency => 'EUR',
      :payment_type => 'cc',
      :invoice_text => 'Payment #1234',
      :options => {
        :payment_id => 1111 # e.g your internal payment handle
      }
    })
  end

  def callback
    # requested from the ipayment servers
    handle_ipayment_callback(@ipayment_config) do |ipayment|
      # ipayment parameter is an Ipayment::Payment instance
      # This is the right place to handle a payment, e.g credit the payment amount to the user.
      # You've access to the options passed to the Ipayment::Payment instance, e.g:
      # ipayment.options['payment_id'] #=> 1111
    end
    head :ok
  end

  def success
    # user gets redirected here if transaction succeeds
    # the options, in this example the payment_id get passed in the params
    # e.g params[:payment_id] #=> 1111
  end

  def error
    # user gets redirected here if transaction fails
    @ipayment = Ipayment::Payment.parse_from_params(@ipayment_config, params)
    if @ipayment && @ipayment.error_code != 0
      @error = @ipayment.error_message
    elsif params[:internal] == 'true'
      @error = "An error occured while processing the transaction"
    end
  end

  protected

    def set_ipayment_config
      @ipayment_config = {
        'accountId' => 99999,
        'trxuserId' =>  99999,
        'trxpassword' => 0,
        'adminactionpassword' => '5cfgRT34xsdedtFLdfHxj7tfwx24fe'
      }
    end
end


example/recurring/subscriptions_controller.rb
class SubscriptionsController
  
  def new
    @ipayment = Ipayment::Payment.new(@ipayment_config_to_use, {
      :amount => 500,
      :currency => 'EUR',
      :payment_type => 'cc',
      :invoice_text => @user.id,
      :options => {
        :user_id => @user.id
      }
    })
  end

  def ipayment_callback
    # this action is not called in the context of the users session
    handle_ipayment_callback(@ipayment_config) do |ipayment|
      if ipayment && ipayment.successful?
        Subscription.create(
          :user_id => ipayment['user_id'],                       # from the options
          :initial_transaction_number => params[:ret_trx_number] # this is the reference used for recurring payments
        )
      else
        # Oops
      end
    end
    head :ok
  end

  def ipayment_success
    session[:subscription_payment_method] = "creditcard"
    session[:initial_transaction_number] = params[:ret_trx_number] 
    if current_user && current_user.credit_card
      redirect_to @@exit_path
    else
      redirect_to payment_data_path
    end
  end

  def ipayment_error
    @ipayment = Ipayment::Payment.parse_from_params(@ipayment_config, params)
    if @ipayment && @ipayment.error_code != 0
      flash[:warning] = @ipayment.error_message
    end
    redirect_to payment_data_path
  end
end


lib/ipayment.rb
# -*- encoding : utf-8 -*-
require 'soap/wsdlDriver'

module Ipayment

  class Connection
    IPAYMENT_WSDL_URL = "https://ipayment.de/v2/ip_service_v3.php?wsdl" unless const_defined?('IPAYMENT_WSDL_URL')
    attr_reader :driver

    def initialize
      @driver = SOAP::WSDLDriverFactory.new(IPAYMENT_WSDL_URL).create_rpc_driver
      @driver.wiredump_file_base = "log/ipayment-log" if Rails.env.development?
    end
  end

  module Helpers
    def form_for_ipayment_auth(ipayment, options = {}, &block)
      ipayment_session = ipayment.generate_session_id({
        :transaction_type => "auth",
        :success_url => options.delete(:success_url),
        :error_url => options.delete(:error_url)
      })

      # extract options we don't want to pass on to the form
      callback_url = options.delete(:callback_url) 
      storage_id   = options.delete(:from_datastorage_id)

      form_tag("https://ipayment.de/merchant/#{ipayment.config['accountId']}/processor/2.0/", options) do
        concat hidden_field_tag(:ipayment_session_id, ipayment_session)
        concat hidden_field_tag(:silent, 1)
        concat hidden_field_tag(:return_paymentdata_details, 1)
        concat hidden_field_tag(:use_datastorage, 1)
        
        concat hidden_field_tag(:hidden_trigger_url, callback_url) if callback_url
        concat hidden_field_tag(:from_datastorage_id, storage_id) if storage_id
        yield
      end
    end
    
    def form_for_ipayment_check_save(ipayment, options, &block)
      ipayment_session = ipayment.generate_session_id({
        :transaction_type => "check_save",
        :success_url => options.delete(:success_url),
        :error_url => options.delete(:error_url)
      })

      # extract options we don't want to pass on to the form
      callback_url = options.delete(:callback_url)

      form_tag("https://ipayment.de/merchant/#{ipayment.config['accountId']}/processor/2.0/", options) do
        concat hidden_field_tag(:ipayment_session_id, ipayment_session)
        concat hidden_field_tag(:silent, 1)
        concat hidden_field_tag(:return_paymentdata_details, 1)
        concat hidden_field_tag(:use_datastorage, 1)
        concat hidden_field_tag(:hidden_trigger_url, callback_url) if callback_url
        yield
      end
    end
  end

  module Controller

    protected
    def handle_ipayment_callback(config, &block)
      if validate_ipayment_request_source
        ipayment = Ipayment::Payment.parse_from_params(config, params)
        yield ipayment
      else
        yield nil
      end
    end

    def validate_ipayment_request_source
      # ipayment_Technik-Handbuch_2008-08.pdf page 42
      return request.post? && (request.env['REMOTE_ADDR'] =~ /^212\.227\.34\.2(18|19|20)$/) == 0
    end
  end

  class Payment
    attr_accessor :amount, :currency, :payment_type, :invoice_text, :initial_recurring, :recurring, :options, :config
    attr_reader :transaction_number, :cc_data, :storage_id, :error_message, :error_code, :successful

    def initialize(configuration, attributes = {})
      attributes.each do |key, value|
        self.send("#{key}=", value) rescue raise "Invalid attribute #{key}"
      end
      self.config = configuration
      @service = Ipayment::Service.new(configuration)
    end

    def generate_session_id(params)
      params = params.merge({
        :amount => amount,
        :currency => currency,
        :payment_type => payment_type,
        :invoice_text => invoice_text,
        :options => options
      })

      params = params.merge({
        :recurring_typ => 'initial',
        :recurring_frequency => 28, # monthly
        :recurring_expiry => 10.years.from_now.strftime("%Y/%m/%d"), # some future date, get's autocorrected to cc expiry date by recurring_allow_expiry_correction
        :recurring_allow_expiry_correction => 1
      }) if initial_recurring?

      params = params.merge({
        :recurring_typ => 'sequencial'
      }) if recurring?

      @service.create_session(params)
    end
    
    # uses SOAP webservice to charge creditcard
    def charge!(params)
      params = params.merge({
        :amount => amount,
        :currency => currency,
        :invoice_text => invoice_text,
        :options => options
      })

      if initial_recurring?
        params = params.merge({
          :recurring_typ => 'initial',
          :recurring_frequency => 28, # monthly
          :recurring_expiry => 10.years.from_now.strftime("%Y/%m/%d"), # some future date, get's autocorrected to cc expiry date by recurring_allow_expiry_correction
          :recurring_allow_expiry_correction => 1
        })
      elsif recurring?
        params = params.merge({
          :recurring_typ => 'sequencial'
        }) 
      end
      result = @service.reauth(params)[0]
      
      if result.status == "SUCCESS"
        @transaction_number = result.successDetails.retTrxNumber
        @successful = true
        return true
      else
        return false
      end
    end
    
    def initial_recurring?
      @initial_recurring || false
    end
    
    def recurring?
      @recurring || false
    end

    def paid?
      @paid || false
    end
    
    def successful?
      @successful || false
    end

    def self.parse_from_params(config, params)
      payment = self.new(config)
      payment.send(:parse_from_params, params)
      payment
    end

    protected

    def parse_from_params(params)
      self.options        = params
      self.amount         = params['trx_amount']
      self.currency       = params['trx_currency']
      self.payment_type   = params['trx_paymenttyp']
      self.invoice_text   = params['invoice_text']
      @transaction_number = params['ret_trx_number']
      @storage_id         = params['storage_id']
      @error_code         = params['ret_errorcode'].to_i
      @error_message      = params['ret_errormsg']
      @error_message.encode!('utf-8', 'iso-8859-1') if @error_message
      @paid               = (params['ret_status'] == "SUCCESS" && params['trx_typ'] == "auth")
      @successful         = params['ret_status'] == "SUCCESS"
      name = params['paydata_cc_cardowner']
      name.encode!('utf-8', 'iso-8859-1') if name
      @cc_data = {
        :number    => params['paydata_cc_number'],
        :card_type => params['paydata_cc_typ'],
        :name      => name
      }
      if params['paydata_cc_expdate']
        @cc_data[:expiry_month] = params['paydata_cc_expdate'][0..1]
        @cc_data[:expiry_year]  = params['paydata_cc_expdate'][2..3]
      end
    end

  end

  class Service
    
    attr_accessor :config

    def initialize(configuration)
      @config = configuration
      @connection = Ipayment::Connection.new
    end

    def reauth(params)
      raise ":orig_trx_number parameter is needed!" unless params[:orig_trx_number]
      driver.reAuthorize(
        self.account_data,
        params[:orig_trx_number],
        self.transaction_data(params),
        params[:options] || {}
      )
    end

    # required params:
    # * :amount
    # * :currency
    # * :transaction_type
    # * :payment_type
    # * :success_url
    # * :error_url
    # optional params:
    # * :options
    # * :invoice_text
    # returns iPayment session id
    def create_session(params)
      driver.createSession(
        self.account_data,
        self.transaction_data(params),
        params[:transaction_type],
        params[:payment_type],
        params[:options] || {},
        { 'redirectUrl' => params[:success_url], 'silentErrorUrl' => params[:error_url] }
      )[0]
    end

    def expire_datastrorage_id(storage_id)
      # We piggyback the expire command on a base check as there seems no way
      # to explicitly ask to expire a record
      driver.basecheck(
        self.account_data,
        { 'storageData' => { 'expireDatastorage' => true, 'fromDataStorageId' => storage_id } },
        { 'trxAmount' => '500', 'trxCurrency' => 'EUR' } # some arbitrary values
      )
    end

    protected

    def driver
      @connection.driver
    end

    def account_data
      {
        'accountId' => @config['accountId'],
        'trxuserId' => @config['trxuserId'],
        'trxpassword' => @config['trxpassword'],
        'adminactionpassword' => @config['adminactionpassword']
      }
    end

    def transaction_data(params)
      transaction_data = { 'trxAmount' => params[:amount], 'trxCurrency' => params[:currency] }
      transaction_data['invoiceText'] = params[:invoice_text] if params[:invoice_text]

      if params[:recurring_typ]
        transaction_data['recurringData'] = { 'recurringTyp' => params[:recurring_typ] }
        transaction_data['recurringData']['recurringFrequency']             = params[:recurring_frequency] if params[:recurring_frequency]
        transaction_data['recurringData']['recurringExpiry']                = params[:recurring_expiry]    if params[:recurring_expiry]
        transaction_data['recurringData']['recurringAllowExpiryCorrection'] = params[:recurring_allow_expiry_correction] if params[:recurring_allow_expiry_correction]
      end

      return transaction_data
    end
  end

end
